### PR TITLE
feat: support customizable retry and timeout settings on the publisher client

### LINKED ISSUE #222: Allow retry and timeout settings on the publisher client
Allow retry and timeout settings on the [Publisher](https://googleapis.dev/python/pubsub/latest/publisher/api/client.html). 

`timeout=` should take `google.api_core.timeout.ExponentialTimeout` and `google.api_core.timeout.ConstantTimeout` (also `LogicalTimeout` in future work).

The goal is be able to set custom retry and custom timeout on the PublisherClient, like this:

```python
custom_retry = api_core.retry.Retry(
    initial=0.250,  # seconds (default: 0.1)
    maximum=90.0,  # seconds (default: 60.0)
    multiplier=1.45,  # default: 1.3
    deadline=300.0,  # seconds (default: 60.0)
    predicate=api_core.retry.if_exception_type(
        api_core.exceptions.Aborted,
        api_core.exceptions.DeadlineExceeded,
        api_core.exceptions.InternalServerError,
        api_core.exceptions.ResourceExhausted,
        api_core.exceptions.ServiceUnavailable,
        api_core.exceptions.Unknown,
        api_core.exceptions.Cancelled,
    ),
)

custom_timeout=api_core.timeout.ExponentialTimeout(
    initial=1.0,  
    maximum=10.0,  
    multiplier=1.0,  
    deadline=300.0,  
)

# Something like this:
publisher = pubsub_v1.PublisherClient(
    publisher_options = pubsub_v1.types.PublisherOptions(
        retry=custom_retry,
        timeout=custom_timeout,
    ),
)
```

Here are the defaults and how they are set currently: https://github.com/googleapis/googleapis/blob/3a935fab757e09c72afd4aa121147a4c97dccc3e/google/pubsub/v1/pubsub_grpc_service_config.json#L42-L62

Remember to update `UPGRADING.md` as well.

### PR DESCRIPTION
**Pull request by plamut**
Closes #222.
Supersedes #239.

This PR adds support for API core timeout classes to the publisher class, replacing the previous `float` timeouts.

**PR checklist:**
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-pubsub/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

**Comment by cguardia** [comment]
@googlebot I consent.

**Review by jimfulton — COMMENTED** [review]
I think the type UNION should admit floats and ints.

I also think the union should be defined once and reused.

There's an obscene amount of repetition in this package.  It was already there, but may be can not add to it. :)

File: google/cloud/pubsub_v1/types.py [hunk 1]
```diff
@@ -37,6 +39,16 @@
 from google.pubsub_v1.types import pubsub as pubsub_gapic_types
 
 
+TimeoutType = Union[
+    None,
```
Comment (jimfulton) [inline google/cloud/pubsub_v1/types.py]: Did we mean to add `None`?  Was this my fault?

This feels like a breaking change.  We didn't accept `None` before.

File: google/pubsub_v1/services/publisher/async_client.py [hunk 2]
```diff
@@ -315,6 +313,9 @@ async def update_topic(
             ),
         )
 
+        if timeout is None or isinstance(timeout, (int, float)):
+            timeout = timeouts.ConstantTimeout(timeout)
```
Comment (jimfulton) [inline google/pubsub_v1/services/publisher/async_client.py:317]: We shouldn't need except for the `None` case, which is new.  I don't think we should accept `None`.  Apologies if `None` came from me.

`int` and `float` are handled here: 

https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/gapic_v1/method.py#L66-L67

The same comment applies to other places where you added this.

Comment (jimfulton) [reply google/pubsub_v1/services/publisher/async_client.py:317]: And, of course, not DRY.

File: google/cloud/pubsub_v1/publisher/_sequencer/base.py [hunk 3]
```diff
@@ -45,18 +47,17 @@ def unpause(self, message):  # pragma: NO COVER
 
     @staticmethod
     @abc.abstractmethod
-    def publish(self, message, retry=None, timeout=None):  # pragma: NO COVER
+    def publish(
+        self, message, retry=None, timeout: gapic_types.TimeoutType = None
```
Comment (jimfulton) [inline google/cloud/pubsub_v1/publisher/_sequencer/base.py:51]: Can we use `DEFAULT` rather than `None`?  The implementation defaults to `DEFAULT`.

**Review by jimfulton — COMMENTED** [review]
Thanks for these changes.

My only complaint is that we added None as a valid timeout, which causes some pain.

**Comment by plamut** [comment]
@jimfulton  I added none, because I noticed that [ConstantTimeout](https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/timeout.py#L71-L82) actually supports it.

**Comment by plamut** [comment]
Flaky samples tests, 500 internal.

File: google/cloud/pubsub_v1/publisher/_sequencer/base.py (hunk 3 above)
Comment (plamut) [reply google/cloud/pubsub_v1/publisher/_sequencer/base.py:51]: That makes sense, the ABC should be consistent.

**Comment by jimfulton** [comment]
> @jimfulton I added none, because I noticed that [ConstantTimeout](https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/timeout.py#L71-L82) actually supports it.
> 
> (not saying that explicitly disabling any timeouts is a good idea, but we already support that automaticlally if we allow instances of `ConstantTimeout`)

That's fair, but at least you could argue that we should make people work to disable timeouts.  


BTW, I tried to figure out what the effect of passing `None` to `ConstantTimout` would be and got lost in the web of indirection.  :)  To convince myself that I knew what was going on I'd have to spend some quality time with `pdb`, and I don't have time for that today.

In any case, if we want to accept `None`, I think we should handle that in one place (https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/gapic_v1/method.py#L66-L67) so we don't repeat the logic in many places.

**Comment by plamut** [comment]
I don't see a good reason why somebody wanted to potentially block indefinitely, thus I'm fine with removing support for a plain `None` (one can always pass in `ConstantTimeout(None)`, if they _really_ want to). We can then also get rid of the "convert constant to object" code in the generated code, because number constants are already handled in `api-core`.

**Comment by plamut** [comment]
> BTW, I tried to figure out what the effect of passing `None` to `ConstantTimout` would be and got lost in the web of indirection. :) 

If I read it correctly, a `ConstantTimeout(None)` is used as a RPC method [decorator](https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/gapic_v1/method.py#L132). What `ConstantTimeout` does is just [setting](https://github.com/googleapis/python-api-core/blob/7337c6b123735fb9ae5d0e54b4399275719c020c/google/api_core/timeout.py#L98-L102) the `timeout` argument to the wrapped method to the given value (`None` in this case).

**Review by jimfulton — APPROVED** [review]

**Comment by plamut** [comment]
Note to self - this PR modifies `synth.py`, thus after merging, we also need to update https://github.com/googleapis/python-pubsub/pull/411

cc: @parthea

**Pull request merged**